File: app/src/main/java/com/kalenjohnson/chronoduo/saveimport/DsSav.java
package com.kalenjohnson.chronoduo.saveimport;

import java.io.UnsupportedEncodingException;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Parser for Nintendo DS Chrono Trigger {@code .sav} (64 KB EEPROM) saves,
 * ported byte-for-byte from tools/saves/ds_sav.py. See REPORT.md #7 for the
 * format -- with one correction found validating against the fixture: the
 * 7 character records actually start at DS slot offset 0x484, not the 0x480
 * the table names (everything downstream, from the inventory at 0x724 on,
 * already lines up with 0x484 + 7*0x60 = 0x724).
 */
public final class DsSav {
    public static final int SLOT_SIZE = 0x2800;
    public static final int NUM_SLOTS = 3;
    public static final int IMAGE_SIZE = 65536;

    public static final int ARDS_HEADER_SIZE = 500;
    public static final int ARDS_TOTAL_SIZE = 262644;
    public static final int DESMUME_FOOTER_SIZE = 122;
    private static final byte[] DESMUME_SIGNATURE;

    static {
        try {
            DESMUME_SIGNATURE = "|-DESMUME SAVE-|".getBytes("US-ASCII");
        } catch (UnsupportedEncodingException e) {
            throw new AssertionError(e);
        }
    }

    private static final long SLOT_MAGIC = 0xFEDCBA98L;

    // Corrected vs. REPORT.md #7's table (see class doc): records actually
    // start here.
    private static final int CHAR_BLOCK_OFF = 0x484;
    private static final int CHAR_SIZE = 0x60;
    private static final int NUM_CHARS = 7;

    private static final int FLAGS_OFF = 0x250;
    private static final int FLAGS_SIZE = 0x200;
    private static final int SCENE_OFF = 0x450;
    private static final int SCENE_SIZE = 0x30;
    private static final int REGION_A_OFF = 0x050;
    private static final int REGION_A_SIZE = 0x200;

    private static final int TECH_BLOCK_OFF = 0xC90;
    private static final int TECH_BLOCK_SIZE = 45;

    private static final int NAMES_OFF = 0xCC0;
    private static final int NAME_STRIDE = 16;
    private static final int NUM_NAMES = 10;

    private static final int PARTY_OFF = 0xD60;
    private static final int GOLD_OFF = 0xD6C;
    private static final int SAVE_COUNT_OFF = 0xD70;
    private static final int PLAY_TIME_OFF = 0xD74;
    private static final int LOCATION_NAME_ID_OFF = 0xD80;
    private static final int ERA_MASK_OFF = 0xD82;

    /** (section name, capacity, expected category nibble), in stream order. */
    public static final class DsInventorySection {
        public final String name;
        public final int capacity;
        public final int category;

        DsInventorySection(String name, int capacity, int category) {
            this.name = name;
            this.capacity = capacity;
            this.category = category;
        }
    }

    public static final DsInventorySection[] INVENTORY_SECTIONS = {
            new DsInventorySection("weapons", 111, 0x0),
            new DsInventorySection("armour", 50, 0x1),
            new DsInventorySection("helmets", 39, 0x2),
            new DsInventorySection("accessories", 59, 0x3),
            new DsInventorySection("consumables", 43, 0x4),
            new DsInventorySection("key_items", 45, 0x5),
    };

    private DsSav() {
    }

    /** 16-byte fixed Shift-JIS string, NUL-padded; full-width -> ASCII via NFKC. */
    static String decodeDsName(byte[] raw, int off, int len) {
        int n = 0;
        while (n < len && raw[off + n] != 0) n++;
        String s;
        try {
            s = new String(raw, off, n, "Shift_JIS");
        } catch (UnsupportedEncodingException e) {
            throw new AssertionError(e);
        }
        return Normalizer.normalize(s, Normalizer.Form.NFKC);
    }

    public static final class DsChar {
        public final byte[] raw;
        public final long id;
        public final int curHp, maxHp, curMp, maxMp, baseMaxHp;
        public final int power, stamina, speed, magic, hit, evade, mdef, level;
        public final int extraU16;
        public final long exp;
        public final long tpNext;
        public final int equipWeapon, equipArmor, equipHelmet, equipAccessory;
        public final int expNext, tpRelated;
        public final byte[] growth; // 6 bytes
        public final byte[] curStats; // 9 bytes
        public final byte[] perCharConstants; // 4 bytes

        private DsChar(byte[] raw, long id, int curHp, int maxHp, int curMp, int maxMp, int baseMaxHp,
                       int power, int stamina, int speed, int magic, int hit, int evade, int mdef, int level,
                       int extraU16, long exp, long tpNext,
                       int equipWeapon, int equipArmor, int equipHelmet, int equipAccessory,
                       int expNext, int tpRelated, byte[] growth, byte[] curStats, byte[] perCharConstants) {
            this.raw = raw;
            this.id = id;
            this.curHp = curHp; this.maxHp = maxHp; this.curMp = curMp; this.maxMp = maxMp;
            this.baseMaxHp = baseMaxHp;
            this.power = power; this.stamina = stamina; this.speed = speed; this.magic = magic;
            this.hit = hit; this.evade = evade; this.mdef = mdef; this.level = level;
            this.extraU16 = extraU16;
            this.exp = exp; this.tpNext = tpNext;
            this.equipWeapon = equipWeapon; this.equipArmor = equipArmor;
            this.equipHelmet = equipHelmet; this.equipAccessory = equipAccessory;
            this.expNext = expNext; this.tpRelated = tpRelated;
            this.growth = growth; this.curStats = curStats; this.perCharConstants = perCharConstants;
        }

        static DsChar parse(byte[] raw) {
            if (raw.length != CHAR_SIZE) throw new IllegalArgumentException("bad DS char record size");
            int maxHp = u16(raw, 0x08), curHp = u16(raw, 0x0A), maxMp = u16(raw, 0x0C), curMp = u16(raw, 0x0E);
            int baseMaxHp = u16(raw, 0x10);
            int power = raw[0x12] & 0xFF, stamina = raw[0x13] & 0xFF, speed = raw[0x14] & 0xFF, magic = raw[0x15] & 0xFF;
            int hit = raw[0x16] & 0xFF, evade = raw[0x17] & 0xFF, mdef = raw[0x18] & 0xFF, level = raw[0x19] & 0xFF;
            int extraU16 = u16(raw, 0x1A);
            long exp = u32(raw, 0x1C);
            long tpNext = u32(raw, 0x20);
            int eqW = u16(raw, 0x36), eqA = u16(raw, 0x38), eqH = u16(raw, 0x3A), eqAcc = u16(raw, 0x3C);
            int expNext = u16(raw, 0x3E);
            int tpRelated = u16(raw, 0x40);
            byte[] growth = slice(raw, 0x42, 6);
            byte[] curStats = slice(raw, 0x48, 9);
            byte[] perChar = slice(raw, 0x51, 4);
            return new DsChar(raw.clone(), u32(raw, 0x00),
                    curHp, maxHp, curMp, maxMp, baseMaxHp,
                    power, stamina, speed, magic, hit, evade, mdef, level,
                    extraU16, exp, tpNext, eqW, eqA, eqH, eqAcc, expNext, tpRelated,
                    growth, curStats, perChar);
        }
    }

    public static final class DsSlot {
        public final byte[] raw;
        public final boolean used;
        public final boolean lastSaved;
        public final byte[] regionA;
        public final byte[] flags;
        public final byte[] sceneBlock;
        public final DsChar[] chars;
        public final Map<String, List<CtSave.Entry>> inventory;
        public final byte[] techBlock;
        public final String[] names;
        public final byte[] party;
        public final byte[] reserve;
        public final int recruitedMask;
        public final long gold;
        public final long saveCount;
        public final long playTimeSeconds;
        public final int locationNameId;
        public final int eraMask;

        private DsSlot(byte[] raw, boolean used, boolean lastSaved, byte[] regionA, byte[] flags, byte[] sceneBlock,
                       DsChar[] chars, Map<String, List<CtSave.Entry>> inventory, byte[] techBlock, String[] names,
                       byte[] party, byte[] reserve, int recruitedMask, long gold, long saveCount,
                       long playTimeSeconds, int locationNameId, int eraMask) {
            this.raw = raw;
            this.used = used;
            this.lastSaved = lastSaved;
            this.regionA = regionA;
            this.flags = flags;
            this.sceneBlock = sceneBlock;
            this.chars = chars;
            this.inventory = inventory;
            this.techBlock = techBlock;
            this.names = names;
            this.party = party;
            this.reserve = reserve;
            this.recruitedMask = recruitedMask;
            this.gold = gold;
            this.saveCount = saveCount;
            this.playTimeSeconds = playTimeSeconds;
            this.locationNameId = locationNameId;
            this.eraMask = eraMask;
        }

        static DsSlot parse(byte[] raw) {
            if (raw.length != SLOT_SIZE) throw new IllegalArgumentException("bad DS slot size");
            long magic = u32(raw, 0);
            boolean lastSaved = magic == SLOT_MAGIC;

            byte[] regionA = slice(raw, REGION_A_OFF, REGION_A_SIZE);
            byte[] flags = slice(raw, FLAGS_OFF, FLAGS_SIZE);
            byte[] sceneBlock = slice(raw, SCENE_OFF, SCENE_SIZE);

            DsChar[] chars = new DsChar[NUM_CHARS];
            for (int i = 0; i < NUM_CHARS; i++) {
                chars[i] = DsChar.parse(slice(raw, CHAR_BLOCK_OFF + i * CHAR_SIZE, CHAR_SIZE));
            }
            // Unused slots are 0xFF-filled (not zero-filled); the FEDCBA98
            // magic only marks the *last-saved* slot, not "used" (other used
            // slots are 0xFFFFFFFF there too) -- REPORT.md #7 /
            // chrono-trigger.18311.duc.
            boolean used = chars[0].maxHp != 0 && chars[0].maxHp != 0xFFFF && (raw[PARTY_OFF] & 0xFF) != 0xFF;

            int off = CHAR_BLOCK_OFF + NUM_CHARS * CHAR_SIZE;
            if (off != 0x724) throw new IllegalStateException("unexpected inventory offset " + Integer.toHexString(off));
            Map<String, List<CtSave.Entry>> inventory = new LinkedHashMap<>();
            for (DsInventorySection sec : INVENTORY_SECTIONS) {
                List<CtSave.Entry> entries = new ArrayList<>(sec.capacity);
                for (int i = 0; i < sec.capacity; i++) {
                    int packed = u16(raw, off);
                    int count = u16(raw, off + 2);
                    off += 4;
                    int idx = packed & 0xFFF;
                    int cat = packed >>> 12;
                    // 0xFFFF shows up throughout an unused, 0xFF-filled slot
                    // -- not a real category byte, don't validate it.
                    if (packed != 0 && packed != 0xFFFF && cat != sec.category) {
                        throw new IllegalArgumentException(String.format(
                                "unexpected category 0x%X in DS section %s (expected 0x%X)",
                                cat, sec.name, sec.category));
                    }
                    entries.add(new CtSave.Entry(idx, count));
                }
                inventory.put(sec.name, entries);
            }
            if (off != TECH_BLOCK_OFF) throw new IllegalStateException("unexpected tech block offset " + Integer.toHexString(off));

            byte[] techBlock = slice(raw, TECH_BLOCK_OFF, TECH_BLOCK_SIZE);

            String[] names = new String[NUM_NAMES];
            for (int i = 0; i < NUM_NAMES; i++) {
                names[i] = decodeDsName(raw, NAMES_OFF + i * NAME_STRIDE, NAME_STRIDE);
            }

            byte[] party = slice(raw, PARTY_OFF, 3);
            byte[] reserve = slice(raw, PARTY_OFF + 3, 6);
            int recruitedMask = raw[PARTY_OFF + 9] & 0xFF;

            long gold = u32(raw, GOLD_OFF);
            long saveCount = u32(raw, SAVE_COUNT_OFF);
            long playTimeSeconds = u32(raw, PLAY_TIME_OFF);
            int locationNameId = u16(raw, LOCATION_NAME_ID_OFF);
            int eraMask = u16(raw, ERA_MASK_OFF);

            return new DsSlot(raw.clone(), used, lastSaved, regionA, flags, sceneBlock, chars, inventory,
                    techBlock, names, party, reserve, recruitedMask, gold, saveCount, playTimeSeconds,
                    locationNameId, eraMask);
        }
    }

    /**
     * Detects and strips the wrapper around a DS save image, returning the
     * raw 65536-byte EEPROM image. Throws IllegalArgumentException if the
     * data doesn't match any known wrapper (REPORT.md #7).
     */
    public static byte[] stripWrapper(byte[] data) {
        if (data.length == ARDS_TOTAL_SIZE || startsWith(data, "ARDS")) {
            if (data.length < ARDS_HEADER_SIZE) {
                throw new IllegalArgumentException("file looks like an ARDS export but is only " + data.length + " bytes");
            }
            int imageLen = data.length - ARDS_HEADER_SIZE;
            if (imageLen < IMAGE_SIZE) {
                throw new IllegalArgumentException("ARDS export image is " + imageLen + " bytes, expected >= " + IMAGE_SIZE);
            }
            byte[] image = new byte[IMAGE_SIZE];
            System.arraycopy(data, ARDS_HEADER_SIZE, image, 0, IMAGE_SIZE);
            return image;
        }

        if (data.length >= DESMUME_FOOTER_SIZE && containsSignature(data, data.length - DESMUME_FOOTER_SIZE, DESMUME_FOOTER_SIZE)) {
            int imageLen = data.length - DESMUME_FOOTER_SIZE;
            if (imageLen != IMAGE_SIZE) {
                throw new IllegalArgumentException("DeSmuME .dsv image (after footer strip) is " + imageLen
                        + " bytes, expected " + IMAGE_SIZE);
            }
            byte[] image = new byte[IMAGE_SIZE];
            System.arraycopy(data, 0, image, 0, IMAGE_SIZE);
            return image;
        }

        if (data.length == IMAGE_SIZE) {
            return data.clone();
        }

        if (data.length == 524288) {
            byte[] image = new byte[IMAGE_SIZE];
            System.arraycopy(data, 0, image, 0, IMAGE_SIZE);
            return image;
        }

        throw new IllegalArgumentException("unrecognized DS save wrapper: " + data.length + " bytes");
    }

    /** Returns 3 DsSlot, never null (unused slots still parse, with used=false). */
    public static DsSlot[] parseImage(byte[] image) {
        if (image.length != IMAGE_SIZE) throw new IllegalArgumentException("expected 65536-byte DS image");
        DsSlot[] slots = new DsSlot[NUM_SLOTS];
        for (int i = 0; i < NUM_SLOTS; i++) {
            slots[i] = DsSlot.parse(slice(image, i * SLOT_SIZE, SLOT_SIZE));
        }
        return slots;
    }

    public static DsSlot[] parseFile(byte[] data) {
        return parseImage(stripWrapper(data));
    }

    private static boolean startsWith(byte[] data, String prefix) {
        byte[] p;
        try {
            p = prefix.getBytes("US-ASCII");
        } catch (UnsupportedEncodingException e) {
            throw new AssertionError(e);
        }
        if (data.length < p.length) return false;
        for (int i = 0; i < p.length; i++) if (data[i] != p[i]) return false;
        return true;
    }

    private static boolean containsSignature(byte[] data, int from, int len) {
        int end = from + len - DESMUME_SIGNATURE.length;
        outer:
        for (int i = from; i <= end; i++) {
            for (int j = 0; j < DESMUME_SIGNATURE.length; j++) {
                if (data[i + j] != DESMUME_SIGNATURE[j]) continue outer;
            }
            return true;
        }
        return false;
    }

    private static byte[] slice(byte[] a, int off, int len) {
        byte[] out = new byte[len];
        System.arraycopy(a, off, out, 0, len);
        return out;
    }

    private static int u16(byte[] a, int off) {
        return (a[off] & 0xFF) | ((a[off + 1] & 0xFF) << 8);
    }

    private static long u32(byte[] a, int off) {
        return (a[off] & 0xFFL) | ((a[off + 1] & 0xFFL) << 8)
                | ((a[off + 2] & 0xFFL) << 16) | ((a[off + 3] & 0xFFL) << 24);
    }
}

File: tools/saves/ds_sav.py
"""Parser for Nintendo DS Chrono Trigger `.sav` (64 KB EEPROM) saves.

See REPORT.md #7 for the full format spec this module implements. One
correction versus that section, found while validating against the fixture
(`ds/chrono-trigger.22851.dst`): the table lists the character-record block
as starting at DS slot offset 0x480, but in the fixture the 7 records
actually start at **0x484** -- the `u32 id` field (0, 1, .. 6) and the
`80 86 90 xx` per-character constant land at DS-record-relative +0x00/+0x04
exactly as documented, they're just 4 bytes further into the slot than the
table's "DS slot" column says. Everything downstream (inventory at 0x724
onward, tech block at 0xC90, names at 0xCC0, party/gold/time/etc.) already
lines up with base 0x484 + 7*0x60 = 0x724, so only that one start offset was
off; nothing else in #7 needed correcting against this fixture.

Wrappers stripped before parsing (REPORT.md #7):
  - Action Replay DS export: 500-byte `ARDS` header, file size 262644 (or
    starts with b"ARDS"); the header is followed by a 256 KB image of which
    only the first 64 KB is real (the rest is 0xFF padding) -- REPORT.md
    only calls this out explicitly for the 524288-byte case, but the same
    "take the first 65536 bytes" rule applies here too (256 KB != 512 KB,
    a case REPORT.md's wrapper list doesn't enumerate by size).
  - raw 65536 bytes: used as-is.
  - 524288 bytes: first 65536 bytes (same 0xFF-padding pattern).
  - DeSmuME .dsv: 122-byte footer carrying the ASCII signature
    "|-DESMUME SAVE-|"; stripped only when that signature is actually
    present in the tail (never assumed from size alone).
"""
import struct
import sys
import unicodedata
from dataclasses import dataclass, field

SLOT_SIZE = 0x2800
NUM_SLOTS = 3
IMAGE_SIZE = 65536

ARDS_HEADER_SIZE = 500
ARDS_TOTAL_SIZE = 262644
DESMUME_FOOTER_SIZE = 122
DESMUME_SIGNATURE = b"|-DESMUME SAVE-|"

SLOT_MAGIC = 0xFEDCBA98  # the last-saved slot's u32 at offset 0 (little-endian on disk: 98 BA DC FE)

# Corrected vs. REPORT.md #7's table (see module docstring): the 7 character
# records actually start here, not at slot offset 0x480.
CHAR_BLOCK_OFF = 0x484
CHAR_SIZE = 0x60
NUM_CHARS = 7

FLAGS_OFF = 0x250
FLAGS_SIZE = 0x200
SCENE_OFF = 0x450
SCENE_SIZE = 0x30
REGION_A_OFF = 0x050
REGION_A_SIZE = 0x200

TECH_BLOCK_OFF = 0xC90
TECH_BLOCK_SIZE = 45

NAMES_OFF = 0xCC0
NAME_STRIDE = 16
NUM_NAMES = 10

PARTY_OFF = 0xD60  # party(3) + reserve(6) + recruited(1) = 10 bytes
GOLD_OFF = 0xD6C
SAVE_COUNT_OFF = 0xD70
PLAY_TIME_OFF = 0xD74
LOCATION_NAME_ID_OFF = 0xD80
ERA_MASK_OFF = 0xD82

# Six inventory sections, in stream order, immediately after the character
# block (0x724). idx = id & 0xFFF; category = id >> 12, checked against the
# section's expected nibble (same category numbering as the port's
# (category<<12)|idx equipment encoding: weapons=0, armour=1, helmets=2,
# accessories=3, plus consumables=4 and key_items=5 for the two sections
# that never appear as equipment).
INVENTORY_SECTIONS = [
    ("weapons", 111, 0x0),
    ("armour", 50, 0x1),
    ("helmets", 39, 0x2),
    ("accessories", 59, 0x3),
    ("consumables", 43, 0x4),
    ("key_items", 45, 0x5),
]


def _decode_ds_name(raw: bytes) -> str:
    """16-byte fixed Shift-JIS string, NUL-padded; full-width chars (e.g.
    'Ｃｒｏｎｏ') normalized to ASCII via NFKC before returning."""
    raw = raw.split(b"\x00", 1)[0]
    s = raw.decode("shift_jis", errors="replace")
    return unicodedata.normalize("NFKC", s)


@dataclass
class DsChar:
    """One 0x60-byte DS character record (REPORT.md #7)."""
    raw: bytes
    id: int
    cur_hp: int
    max_hp: int
    cur_mp: int
    max_mp: int
    base_max_hp: int
    power: int
    stamina: int
    speed: int
    magic: int
    hit: int
    evade: int
    mdef: int
    level: int
    extra_u16: int  # +0x1A -- not present in the port (REPORT.md #7)
    exp: int
    tp_next: int  # +0x20 u32 -> port +0x1A
    equip_weapon: int
    equip_armor: int
    equip_helmet: int
    equip_accessory: int
    exp_next: int
    tp_related: int
    growth: bytes  # 6 bytes -> port +0x38
    cur_stats: bytes  # 9 bytes -> port +0x3E
    per_char_constants: bytes  # 4 bytes -> port +0x47

    @staticmethod
    def parse(raw: bytes) -> "DsChar":
        assert len(raw) == CHAR_SIZE
        max_hp, cur_hp, max_mp, cur_mp, base_max_hp = struct.unpack_from("<HHHHH", raw, 0x08)
        power, stamina, speed, magic, hit, evade, mdef, level = raw[0x12:0x1A]
        extra_u16 = struct.unpack_from("<H", raw, 0x1A)[0]
        exp = struct.unpack_from("<I", raw, 0x1C)[0]
        tp_next = struct.unpack_from("<I", raw, 0x20)[0]
        eq_w, eq_a, eq_h, eq_acc = struct.unpack_from("<HHHH", raw, 0x36)
        exp_next = struct.unpack_from("<H", raw, 0x3E)[0]
        tp_related = struct.unpack_from("<H", raw, 0x40)[0]
        return DsChar(
            raw=bytes(raw),
            id=struct.unpack_from("<I", raw, 0x00)[0],
            cur_hp=cur_hp, max_hp=max_hp, cur_mp=cur_mp, max_mp=max_mp,
            base_max_hp=base_max_hp,
            power=power, stamina=stamina, speed=speed, magic=magic,
            hit=hit, evade=evade, mdef=mdef, level=level,
            extra_u16=extra_u16,
            exp=exp, tp_next=tp_next,
            equip_weapon=eq_w, equip_armor=eq_a, equip_helmet=eq_h, equip_accessory=eq_acc,
            exp_next=exp_next, tp_related=tp_related,
            growth=bytes(raw[0x42:0x48]),
            cur_stats=bytes(raw[0x48:0x51]),
            per_char_constants=bytes(raw[0x51:0x55]),
        )


@dataclass
class DsSlot:
    raw: bytes  # full 0x2800-byte slot
    used: bool
    is_last_saved: bool
    region_a: bytes  # 0x200 bytes at 0x050
    flags: bytes  # 0x200 bytes at 0x250 (event flags)
    scene_block: bytes  # 0x30 bytes at 0x450
    chars: list  # 7 DsChar, Crono..Magus
    inventory: dict  # section name -> list[(idx, count)]
    tech_block: bytes  # 45 bytes at 0xC90
    names: list  # 10 decoded strings
    party: bytes  # 3 bytes
    reserve: bytes  # 6 bytes
    recruited_mask: int
    gold: int
    save_count: int  # u32 at 0xD70
    play_time_seconds: int
    location_name_id: int
    era_mask: int

    @staticmethod
    def parse(raw: bytes) -> "DsSlot":
        assert len(raw) == SLOT_SIZE
        magic = struct.unpack_from("<I", raw, 0)[0]
        is_last_saved = magic == SLOT_MAGIC

        region_a = raw[REGION_A_OFF:REGION_A_OFF + REGION_A_SIZE]
        flags = raw[FLAGS_OFF:FLAGS_OFF + FLAGS_SIZE]
        scene_block = raw[SCENE_OFF:SCENE_OFF + SCENE_SIZE]

        chars = [
            DsChar.parse(raw[CHAR_BLOCK_OFF + i * CHAR_SIZE: CHAR_BLOCK_OFF + (i + 1) * CHAR_SIZE])
            for i in range(NUM_CHARS)
        ]

        # Unused slots are 0xFF-filled (not zero-filled); e.g. an all-0xFF
        # max HP or a 0xFF party byte 0. The FEDCBA98 magic only marks the
        # *last-saved* slot, not "used" (the other used slots are 0xFFFFFFFF
        # there too) -- REPORT.md #7 / fixture chrono-trigger.18311.duc.
        used = chars[0].max_hp not in (0, 0xFFFF) and raw[PARTY_OFF] != 0xFF

        off = CHAR_BLOCK_OFF + NUM_CHARS * CHAR_SIZE
        assert off == 0x724, f"unexpected inventory offset {off:#x}"
        inventory = {}
        for name, capacity, expected_cat in INVENTORY_SECTIONS:
            entries = []
            for i in range(capacity):
                packed, count = struct.unpack_from("<HH", raw, off)
                off += 4
                idx = packed & 0xFFF
                cat = packed >> 12
                # 0xFFFF (both fields 0xFF) shows up throughout an unused,
                # 0xFF-filled slot (REPORT.md #7 / chrono-trigger.18311.duc
                # slots 1-2) -- not a real category byte, don't validate it.
                if packed not in (0, 0xFFFF) and cat != expected_cat:
                    raise ValueError(
                        f"unexpected category {cat:#x} in DS section {name} "
                        f"(expected {expected_cat:#x}) at slot offset {off - 4:#x}")
                entries.append((idx, count))
            inventory[name] = entries
        assert off == TECH_BLOCK_OFF, f"unexpected tech block offset {off:#x}"

        tech_block = raw[TECH_BLOCK_OFF:TECH_BLOCK_OFF + TECH_BLOCK_SIZE]

        names = [
            _decode_ds_name(raw[NAMES_OFF + i * NAME_STRIDE: NAMES_OFF + (i + 1) * NAME_STRIDE])
            for i in range(NUM_NAMES)
        ]

        party = raw[PARTY_OFF:PARTY_OFF + 3]
        reserve = raw[PARTY_OFF + 3:PARTY_OFF + 9]
        recruited_mask = raw[PARTY_OFF + 9]

        gold = struct.unpack_from("<I", raw, GOLD_OFF)[0]
        save_count = struct.unpack_from("<I", raw, SAVE_COUNT_OFF)[0]
        play_time_seconds = struct.unpack_from("<I", raw, PLAY_TIME_OFF)[0]
        location_name_id = struct.unpack_from("<H", raw, LOCATION_NAME_ID_OFF)[0]
        era_mask = struct.unpack_from("<H", raw, ERA_MASK_OFF)[0]

        return DsSlot(
            raw=bytes(raw), used=used, is_last_saved=is_last_saved,
            region_a=bytes(region_a), flags=bytes(flags), scene_block=bytes(scene_block),
            chars=chars, inventory=inventory, tech_block=bytes(tech_block), names=names,
            party=bytes(party), reserve=bytes(reserve), recruited_mask=recruited_mask,
            gold=gold, save_count=save_count, play_time_seconds=play_time_seconds,
            location_name_id=location_name_id, era_mask=era_mask,
        )


def strip_wrapper(data: bytes) -> bytes:
    """Detect and strip the wrapper around a DS save image, returning the
    raw 65536-byte EEPROM image. Raises ValueError if the file doesn't match
    any known wrapper (REPORT.md #7)."""
    if len(data) == ARDS_TOTAL_SIZE or data[:4] == b"ARDS":
        if len(data) < ARDS_HEADER_SIZE:
            raise ValueError(f"file looks like an ARDS export but is only {len(data)} bytes")
        image = data[ARDS_HEADER_SIZE:]
        if len(image) > IMAGE_SIZE:
            image = image[:IMAGE_SIZE]
        elif len(image) < IMAGE_SIZE:
            raise ValueError(f"ARDS export image is {len(image)} bytes, expected >= {IMAGE_SIZE}")
        return image

    if len(data) >= DESMUME_FOOTER_SIZE and DESMUME_SIGNATURE in data[-DESMUME_FOOTER_SIZE:]:
        image = data[:-DESMUME_FOOTER_SIZE]
        if len(image) != IMAGE_SIZE:
            raise ValueError(f"DeSmuME .dsv image (after footer strip) is {len(image)} bytes, "
                              f"expected {IMAGE_SIZE}")
        return image

    if len(data) == IMAGE_SIZE:
        return data

    if len(data) == 524288:
        return data[:IMAGE_SIZE]

    raise ValueError(
        f"unrecognized DS save wrapper: {len(data)} bytes, "
        f"first 4 bytes {data[:4]!r}")


def parse_ds_image(image: bytes):
    """Return a list of 3 DsSlot (never None -- unused slots are still
    parsed, just with used=False; REPORT.md #7 notes the magic can't be used
    to determine which slots hold data)."""
    assert len(image) == IMAGE_SIZE
    slots = []
    for i in range(NUM_SLOTS):
        raw = image[i * SLOT_SIZE:(i + 1) * SLOT_SIZE]
        slots.append(DsSlot.parse(raw))
    return slots


def load_ds_sav(path: str):
    with open(path, "rb") as f:
        data = f.read()
    image = strip_wrapper(data)
    return parse_ds_image(image)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"=== {path} ===")
        slots = load_ds_sav(path)
        for i, slot in enumerate(slots):
            if not slot.used:
                print(f"  slot {i}: unused")
                continue
            party_ids = [b for b in slot.party if b < 7]
            party_names = [slot.names[cid] for cid in party_ids]
            levels = [slot.chars[cid].level for cid in party_ids]
            hours = slot.play_time_seconds // 3600
            minutes = (slot.play_time_seconds % 3600) // 60
            seconds = slot.play_time_seconds % 60
            print(f"  slot {i}: last_saved={slot.is_last_saved}")
            print(f"    party: {', '.join(party_names)}  levels: {levels}")
            print(f"    gold: {slot.gold}  play time: {hours:02d}:{minutes:02d}:{seconds:02d} "
                  f"({slot.play_time_seconds}s)")
            print(f"    location name id: {slot.location_name_id:#04x}  era mask: {slot.era_mask:#06x}")
            print(f"    recruited mask: {slot.recruited_mask:#04x}  save count: {slot.save_count}")
            print(f"    names: {slot.names}")
            for c in slot.chars:
                name = slot.names[c.id] if c.id < 8 else f"id{c.id}"
                print(f"    {name}: lvl {c.level} hp {c.cur_hp}/{c.max_hp} mp {c.cur_mp}/{c.max_mp} exp {c.exp}")
